[@types/theme-ui] Error on spreading styled system interpolation objects

- [X] I tried using the `@types/xxxx` package and had problems.
- [X] I tried using the latest stable version of tsc. https://www.npmjs.com/package/typescript
- [x] [Mention](https://github.com/blog/821-mention-somebody-they-re-notified) the authors (see `Definitions by:` in `index.d.ts`) so they can respond.

@erikdstock @ifiokjr @sbardian @pomber @wKovacs64 @prateekkathal

---

Hi! I've encountered a type error while trying to compose reusable style objects. 

**TLDR**: Type 'SystemCssProperties' is not assignable to type 'UseThemeFunction'.

### Reproduction

[CodeSandbox](https://codesandbox.io/s/theme-ui-sx-type-error-repro-3uvyb)

```tsx
import { SystemCssProperties } from "@styled-system/css";

const buttonStyles: SystemCssProperties = {
  font: "inherit",
  background: "papayawhip",
  border: "none"
};

const styles = { buttonStyles };

const _ = (
  <button
    type="button"
    sx={styles.buttonStyles} // 🔥
  >
    click me
  </button>
);
```

Typing `buttonStyles` as `SxStyleProp` works, but then you can't spread it onto other objects like

```tsx
  sx={{ ...styles.buttonReset, background: "blue" }}
```

### Errors

For the first snippet:

```
error TS2322: Type 'SystemCssProperties' is not assignable to type '(SystemCssProperties & Record<string, string | number | SystemCssProperties | CSSPseudoSelectorProps | ... 4 
more ... | Record<...>>) | ... 4 more ... | undefined'.
  Type 'SystemCssProperties' is not assignable to type 'UseThemeFunction & Record<string, string | number | 
SystemCssProperties | CSSPseudoSelectorProps | CSSSelectorObject | VariantProperty | UseThemeFunction | (string | ... 1 more ... | null)[] | Record<...>>'.       
    Type 'SystemCssProperties' is not assignable to type 'UseThemeFunction'.
      Type 'SystemCssProperties' provides no match for the signature '(theme: any): SystemStyleObject'.     

165   <button type="button" sx={styles.buttonStyles}> 
                            ~~

  node_modules/@types/theme-ui/index.d.ts:157:5       
    157     sx?: SxStyleProp;
            ~~
    The expected type comes from property 'sx' which is declared here on type 'DetailedHTMLProps<ButtonHTMLAttributes<HTMLButtonElement>, HTMLButtonElement>'
```

and for the second one:
```tsx
Type '{ background: string; clipPath?: string | SystemCssProperties | CSSPseudoSelectorProps | CSSSelectorObject | ... 4 more ... | undefined; ... 428 more ...; fontWeight?: string | ... 8 more ... | undefined; }' is not assignable to type '(SystemCssProperties & Record<string, string | number | SystemCssProperties | CSSPseudoSelectorProps | ... 4 more ... | Record<...>>) | ... 4 more ... | undefined'.
  Type '{ background: string; clipPath?: string | SystemCssProperties | CSSPseudoSelectorProps | CSSSelectorObject | ... 4 more ... | undefined; ... 428 more ...; fontWeight?: string | ... 8 more ... | undefined; }' is not assignable to type 'UseThemeFunction & Record<string, string | number | SystemCssProperties | CSSPseudoSelectorProps | CSSSelectorObject | VariantProperty | UseThemeFunction | (string | ... 1 more ... | null)[] | Record<...>>'.
    Type '{ background: string; clipPath?: string | SystemCssProperties | CSSPseudoSelectorProps | CSSSelectorObject | ... 4 more ... | undefined; ... 428 more ...; fontWeight?: string | ... 8 more ... | undefined; }' is not assignable to type 'UseThemeFunction'.
      Type '{ background: string; clipPath?: string | SystemCssProperties | CSSPseudoSelectorProps | CSSSelectorObject | ... 4 more ... | undefined; ... 428 more ...; fontWeight?: string | ... 8 more ... | undefined; }' provides no match for the signature '(theme: any): SystemStyleObject'.ts(2322)
```

### A workaround
 
It "fixed" it locally with a small change to SxStyleProp definition.

```tsx
export type SxStyleProp =  SystemStyleObject | (SystemStyleObject &
    Record<
        string,
        | SystemStyleObject
        | ResponsiveStyleValue<number | string>
        | Record<string, SystemStyleObject | ResponsiveStyleValue<number | string>>
    >);
```

### Problem

The workaround fails at runtime when extra properties don't match what styled-system/css expects.

Example:
```ts
const badSystemStyleObject: SystemStyleObject = {
  background: "blue",
  foo(_arg1: unknown, arg2: { bar: string }): string {
    // This will raise TypeError: Cannot read property 'bar' of undefined on
    // https://github.com/styled-system/styled-system/blob/master/packages/css/src/index.js#L191
    return arg2.bar;
  }
};
```

--- a/types/theme-ui/index.d.ts
+++ b/types/theme-ui/index.d.ts
@@ -123,17 +123,11 @@ export interface Theme extends StyledSystemTheme {
 export const jsx: typeof React.createElement;
 
 /**
- * The `SxStyleProp` extension `SystemStyleObject` and `Emotion` [style props](https://emotion.sh/docs/object-styles)
- * such that properties that are part of the `Theme` will be transformed to
- * their corresponding values. Other valid CSS properties are also allowed.
+ * The `sx` prop accepts a `SxStyleProp` object and properties that are part of
+ * the `Theme` will be transformed to their corresponding values. Other valid
+ * CSS properties are also allowed.
  */
-export type SxStyleProp = SystemStyleObject &
-    Record<
-        string,
-        | SystemStyleObject
-        | ResponsiveStyleValue<number | string>
-        | Record<string, SystemStyleObject | ResponsiveStyleValue<number | string>>
-    >;
+export type SxStyleProp = SystemStyleObject;
 
 export interface SxProps {
     /**

--- a/types/theme-ui/theme-ui-tests.tsx
+++ b/types/theme-ui/theme-ui-tests.tsx
@@ -1,5 +1,5 @@
 /** @jsx jsx */
-import { Flex, jsx, InitializeColorMode, ColorMode, Styled, Theme } from 'theme-ui';
+import { Flex, jsx, InitializeColorMode, ColorMode, Styled, SxStyleProp, Theme } from 'theme-ui';
 
 export const Component = () => {
     return (
@@ -112,3 +112,25 @@ const themeWithStyles: Theme = {
         },
     },
 };
+
+function SpreadingAndMergingInSxProp() {
+    const buttonStyles: SxStyleProp = {
+        font: 'inherit',
+        color: 'primary',
+        background: 'papayawhip',
+        border: 'none',
+        boxShadow: () => `0 0 4px black`,
+        ':hover': {
+            background: 'yellow',
+            width: ['80%', '40%', '20%'],
+        },
+        height: '2rem',
+        width: ['100%', '50%', '25%'],
+    };
+
+    return (
+        <button type="button" sx={{ ...buttonStyles, background: 'red' }}>
+            click me
+        </button>
+    );
+}

